File: NeetCode/Math_And_Geometry/Detect_Squares/detect__squares.py
from typing import Optional, List
from collections import defaultdict


class CountSquares:

    def __init__(self):
        self.ptCount = defaultdict(int)
        self.points = []

    def add(self, point: List[int]) -> None:
        self.ptCount[tuple(point)] += 1
        self.points.append(point)

    def count(self, point: List[int]) -> int:
        res = 0
        px, py = point
        for x, y in self.points:
            if (abs(py - y) != abs(px - x)) or x == px or y == py:
                continue
            res += self.ptCount[(x, py)] * self.ptCount[(px, y)]
        return res


if __name__ == "__main__":
    sol = CountSquares()

    sol.add([1, 1])
    sol.add([2, 2])
    sol.add([1, 2])

    print(sol.count([2, 1]))
    print("Running Solution...")

File: NeetCode/Math_And_Geometry/Detect_Squares/Detect_Squares.cpp
// Detect_Squares.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class CountSquares
{
private:
    struct PairHash
    {
        size_t operator()(const pair<int, int> &p) const
        {
            return hash<int>()(p.first) ^ (hash<int>()(p.second) << 1);
        }
    };

    unordered_map<pair<int, int>, int, PairHash> ptCount;
    vector<pair<int, int>> points;

public:
    CountSquares()
    {
    }

    void add(vector<int> point)
    {
        pair<int, int> pt = {point[0], point[1]};
        ptCount[pt]++;
        points.push_back(pt);
    }

    int count(vector<int> point)
    {
        int result = 0;
        int py = point[1];
        int px = point[0];
        for (pair<int, int> xy : points)
        {
            int x = xy.first;
            int y = xy.second;
            if (abs(py - y) != abs(px - x) || x == px || y == py)
            {
                continue;
            }
            result = result + ptCount[{x, py}] * ptCount[{px, y}];
        }
        return result;
    }
};
